utilities/object_registry.cc

```diff
@@ -26,61 +55,412 @@ const ObjectLibrary::Entry *ObjectLibrary::FindEntry(
+                           const std::string &method,
+                           const std::string &arg) override;
+
+  // Returns the number of registered types for this registry.
+  // If specified (not-null), types is updated to include the names of the
+  // registered types.
+  size_t GetRegisteredTypes(
+      std::unordered_set<std::string> *types) const override {
+    assert(types);
+    for (const auto &library : options_.libraries) {
+      library->GetRegisteredTypes(types);
+    }
+    return types->size();
+  }
+
+  // Returns the number of registered names for the input type
+  // If specified (not-null), names is updated to include the names for the type
+  size_t GetRegisteredNames(const std::string &type) const override {
+    return GetRegisteredNames(type, nullptr);
+  }
+
+  size_t GetRegisteredNames(const std::string &type,
+                            std::vector<std::string> *names) const override {
+    size_t count = 0;
+    for (const auto &library : options_.libraries) {
+      count += library->GetRegisteredNames(type, names);
+    }
+    return count;
+  }
+
+  // Returns the total number of factories registered for this library.
+  // This method returns the sum of all factories registered for all types.
+  // @param num_types returns how many unique types are registered.
+  size_t GetFactoryCount(size_t *num_types) const override {
+    std::unordered_set<std::string> types;
+    *num_types = GetRegisteredTypes(&types);
+    size_t count = 0;
+    for (const auto &type : types) {
+      count += GetRegisteredNames(type, nullptr);
+    }
+    return count;
+  }
+
+  Status PrepareOptions(const ConfigOptions &config_options) override {
+    for (const auto &library : options_.libraries) {
+      Status s = library->PrepareOptions(config_options);
+      if (!s.ok()) {
+        return s;
+      }
+    }
+    return Configurable::PrepareOptions(config_options);
+  }
+
+ protected:
+  // Searches (from back to front) the libraries looking for the
+  // an entry that matches this pattern.
+  // Returns the entry if it is found, and nullptr otherwise
+  const ObjectLibrary::Entry *FindEntry(
+      const std::string &type, const std::string &name) const override {
+    for (auto iter = options_.libraries.crbegin();
+         iter != options_.libraries.crend(); ++iter) {
+      const auto *entry = iter->get()->FindEntry(type, name);
+      if (entry != nullptr) {
+        return entry;
+      }
+    }
+    return nullptr;
+  }
+
+  Status ConfigureOptions(
+      const ConfigOptions &config_options,
+      const std::unordered_map<std::string, std::string> &opt_map,
+      std::unordered_map<std::string, std::string> *unused) override {
+    // We want to control when any libraries are initialized.
+    // Turn off prepare while we are configuring the registry.
+    // Once the registry is completely configured, we can go back and configure
+    // the libraries.
+    ConfigOptions copy = config_options;
```

Comment:
I several places, the variable name "copy" and "left" are used. Maybe replacing with more explicit name.